purchase/app/purchase/CheckoutAddressController.java

```diff
@@ -0,0 +1,105 @@
+
+    @Inject
+    public CheckoutAddressController(final ControllerDependency controllerDependency, final ProductDataConfig productDataConfig) {
+        super(controllerDependency);
+        this.productDataConfig = productDataConfig;
+    }
+
+    public F.Promise<Result> show(final String languageTag) {
+        final UserContext userContext = userContext(languageTag);
+        final F.Promise<Cart> cartPromise = getOrCreateCart(userContext, session());
+        return cartPromise.map(cart -> {
+            final Messages messages = messages(userContext);
+            final CheckoutAddressPageContent content = new CheckoutAddressPageContent(cart, messages, configuration(), userContext, productDataConfig, reverseRouter());
+            final SunrisePageData pageData = pageData(userContext, content, ctx());
+            return ok(templateService().renderToHtml("checkout-address", pageData, userContext.locales()));
+        });
+    }
+
+    @AddCSRFToken
+    @RequireCSRFCheck
+    public F.Promise<Result> process(final String languageTag) {
+        final UserContext userContext = userContext(languageTag);
+        return getOrCreateCart(userContext, session()).flatMap(cart -> {
+            final CheckoutAddressFormData checkoutAddressFormData = extractBean(request(), CheckoutAddressFormData.class);
+            final Form<CheckoutAddressFormData> filledForm = obtainFilledForm(checkoutAddressFormData);
+            final Messages messages = messages(userContext);
+            final CheckoutAddressPageContent content = new CheckoutAddressPageContent(checkoutAddressFormData, cart, messages, configuration(), userContext, productDataConfig, reverseRouter());
+            if (filledForm.hasErrors()) {
+                return F.Promise.pure(badRequest(userContext, filledForm, content));
+            } else {
+                return updateCart(sphere(), cart, content)
+                        .map(updatedCart -> redirect(reverseRouter().showCheckoutShippingForm(languageTag)));
+            }
+        });
+    }
+
+    private F.Promise<Cart> updateCart(final PlayJavaSphereClient sphere, final Cart cart, final CheckoutAddressPageContent content) {
+        final Address shippingAddress = content.getAddressForm().getShippingAddress().toAddress();
+        final Address nullableBillingAddress = content.getAddressForm().isBillingAddressDifferentToBillingAddress()
+                ? content.getAddressForm().getBillingAddress().toAddress()
+                : null;
+        final List<UpdateAction<Cart>> updateActions = asList(
+                SetShippingAddress.of(shippingAddress),
+                SetBillingAddress.of(nullableBillingAddress)
+        );
+        return sphere.execute(CartUpdateCommand.of(cart, updateActions));
+    }
+
+    private Result badRequest(final UserContext userContext, final Form<CheckoutAddressFormData> filledForm, final CheckoutAddressPageContent content) {
+        Logger.info("cart not valid");
```

Comment:
I'll take it into account on the clean up phase, this was just creating the new step out of the old step.